lib/mayaHydra/hydraExtensions/adapters/materialAdapter.cpp

```diff
@@ -231,10 +246,81 @@ class MayaHydraShadingEngineAdapter : public MayaHydraMaterialAdapter
         }
     }
 
+    bool PopulateMaterialXNetworkMap(HdMaterialNetworkMap& networkMap)
+    {
+        // Get the dependency node
+        MStatus status;
+        MFnDependencyNode node(_surfaceShader, &status);
+        if (!status) {
+            return false;
+        }
+
+        // Fetch the "renderDocument" attribute from the node
+        static const MString renderDocumentStr("renderDocument");  
+        auto mtlxDocPlug = node.findPlug(renderDocumentStr, true, &status);
+        if (!status) {
+            return false;
+        }
+
+        // Construct a MaterialX document
+        auto mtlxDocStr = mtlxDocPlug.asString();
+        auto mtlxDoc = MaterialX::createDocument();
+        MaterialX::readFromXmlString(mtlxDoc, mtlxDocStr.asChar());
+
+        // Create a Usd Stage from the MaterialX document
+        auto stage = UsdStage::CreateInMemory("tmp.usda", TfNullPtr);
+        UsdMtlxRead(mtlxDoc, stage);
+        
+        // Search for material group in the Usd Stage
+        static const SdfPath basePath("/MaterialX/Materials");
+        auto mtlxRange = stage->GetPrimAtPath(basePath).GetChildren();
+        if (mtlxRange.empty()) {
+            return false;
+        }
+
+        // There should be only one material. Fetch it.
+        UsdShadeMaterial mtlxMaterial(*mtlxRange.begin());
+        if (!mtlxMaterial) {
+            return false;
+        }
+
+        // Get MaterialX output
+        UsdShadeOutput mtlxOutput = mtlxMaterial.GetOutput(_tokens->mtlxSurface);
+        if (!mtlxOutput) {
+            return false;
+        }
+
+        // Get MaterialX shader outputs
+        UsdShadeAttributeVector mtlxShaderOutputs = 
+            UsdShadeUtils::GetValueProducingAttributes(mtlxOutput, /*shaderOutputsOnly*/true);
+        if (mtlxShaderOutputs.empty()) {
+            return false;
+        }
+        
+        // Finally get MaterialX shader
+        UsdShadeShader mtlxShader(mtlxShaderOutputs[0].GetPrim());
+        if (!mtlxShader) {
+            return false;
+        }
+
+        // Convert the MaterialX shader to HdMaterialNetwork
+        UsdImagingBuildHdMaterialNetworkFromTerminal(
+            mtlxShader.GetPrim(), _tokens->surface, {_tokens->mtlx},
+            {_tokens->mtlx}, &networkMap, UsdTimeCode());
+            
+        return true;
+    }
+
     VtValue GetMaterialResource() override
     {
         TF_DEBUG(MAYAHYDRALIB_ADAPTER_MATERIALS)
             .Msg("MayaHydraShadingEngineAdapter::GetMaterialResource(): %s\n", GetID().GetText());
+        
+        HdMaterialNetworkMap materialXNetworkMap;
+        if (PopulateMaterialXNetworkMap(materialXNetworkMap)) {
```

Comment:
is there any reason not to do the translation inside MayaHydraMaterialNetworkConverter? I feel it's better to move the code into MayaHydraMaterialNetworkConverter as MatX is also one kind of Material.

Reply:
The interface of MayaHydraMaterialNetworkConverter seems to be developped specifically for UsdPreviewSurface network. I think we would have hard times adapting and rewriting it for the needs of MaterialX. For example method HdMaterialNode* GetMaterial would need to change its output type as well as all other public methods have no meaning for MaterialX (like GetPreviewShaderParams). I don't see any reason for readapting  MayaHydraMaterialNetworkConverter for MaterialX

Reply:
ok, sounds like having a base materialConverter and diferent sub materialConverters would make the architecture looks better, as the actual conversion is the responsibility of materialConverter, not materialAdapter.

I'm fine with the current code, just trying to make it easier to maintain,  we could also refactor it later when we need more materialConverters in the future.